core/drivers/crypto/caam/hash/caam_hash.c

```diff
@@ -0,0 +1,636 @@
+		caam_desc_add_word(desc, FIFO_LD_EXT(CLASS_2, MSG, LAST_C2));
+		caam_desc_add_ptr(desc, paddr_data);
+		caam_desc_add_word(desc, size_inmade);
+
+		/* Clean the input data to be loaded */
+		cache_operation(TEE_CACHECLEAN, (void *)data, size_inmade);
+
+		/* Save the running context */
+		caam_desc_add_word(desc, ST_NOIMM(CLASS_2, REG_CTX,
+						  hashdata->ctx.length));
+		caam_desc_add_ptr(desc, hashdata->ctx.paddr);
+
+		HASH_DUMPDESC(desc);
+
+		/* Ensure Context register data are not in cache */
+		cache_operation(TEE_CACHEINVALIDATE, hashdata->ctx.data,
+				hashdata->ctx.length);
+
+		jobctx.desc = desc;
+		retstatus = caam_jr_enqueue(&jobctx, NULL);
+
+		if (retstatus == CAAM_NO_ERROR) {
+			ret = TEE_SUCCESS;
+			HASH_DUMPBUF("CTX", hashdata->ctx.data,
+				     hashdata->ctx.length);
+		} else {
+			HASH_TRACE("CAAM Status 0x%08" PRIx32, jobctx.status);
+			ret = job_status_to_tee_result(jobctx.status);
+		}
+	} else {
+		ret = TEE_SUCCESS;
+
+		if (size_topost) {
+			/* All input data must be saved */
+			size_inmade = 0;
+		}
+	}
+
+	if (size_topost && data) {
+		struct caambuf indata = {
+			.data = (uint8_t *)data,
+			.length = inlen
+		};
+
+		HASH_TRACE("Post %zu of input len %zu made %zu", size_topost,
+			   len, size_inmade);
+		ret = caam_cpy_block_src(&hashdata->blockbuf, &indata,
+					 size_inmade);
+	}
+
+exit_update:
+	if (ret != TEE_SUCCESS)
+		do_free_intern(hashdata);
+
+	return ret;
+}
+
+/*
+ * Finalize the Hash operation
+ *
+ * @ctx     Operation Software context
+ * @digest  [out] Hash digest buffer
+ * @len     Digest buffer length
+ */
+static TEE_Result do_final(struct crypto_hash_ctx *ctx, uint8_t *digest,
+			   size_t len)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	enum caam_status retstatus = CAAM_FAILURE;
+	struct crypto_hash *hash = to_hash_ctx(ctx);
+	struct hashdata *hashdata = hash->ctx;
+	const struct hashalg *alg = NULL;
+	struct caam_jobctx jobctx = {};
+	uint32_t *desc = NULL;
+	int realloc = 0;
+	struct caambuf digest_align = {};
+
+	HASH_TRACE("Hash Final (0x%p)", hashdata);
+
+	if (!digest || !len || !hashdata)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	alg = hashdata->alg;
+
+	if (!hashdata->ctx.data) {
+		retstatus = do_allocate_intern(hashdata);
+		if (retstatus != CAAM_NO_ERROR) {
+			ret = TEE_ERROR_OUT_OF_MEMORY;
+			goto exit_final;
+		}
+	}
+
+	if (alg->size_digest > len) {
+		HASH_TRACE("Digest buffer size %d too short (%zu)",
+			   alg->size_digest, len);
+
+		retstatus =
+			caam_alloc_align_buf(&digest_align, alg->size_digest);
+		if (retstatus != CAAM_NO_ERROR) {
+			HASH_TRACE("Hash digest reallocation error");
+			ret = TEE_ERROR_OUT_OF_MEMORY;
+			goto exit_final;
+		}
+		realloc = 1;
+	} else {
+		realloc =
+			caam_set_or_alloc_align_buf(digest, &digest_align, len);
+
+		if (realloc == -1) {
+			HASH_TRACE("Hash digest reallocation error");
+			ret = TEE_ERROR_OUT_OF_MEMORY;
+			goto exit_final;
+		}
+	}
+
+	HASH_TRACE("Final Type 0x%X - Digest @0x%p-%zu", alg->type,
```

Comment:
s/%X/%"PRIx32"/

minor: remove `0x` in front of `%p`.

Reply:
ok